Use one-shot static hash methods

martincostello opened the pull request:
**PR Title**

Use one-shot static hash methods

**PR Description**

* Use one-shot static `HashData()` methods for `SHA1` and `HMACSHA*`.
* Fix or update some documentation links.
* Expand the test coverage for the `TwitterHandler` class based on the Facebook handler's tests.
* Minor readability change to use `Convert.ToHexString()` in a test helper.

martincostello committed:
Use one-shot static hash methods

Use one-shot static hash methods for SHA1 and HMACSHA*.
Also a readability change to use Convert.ToHexString().

martincostello commented on src/Security/Authentication/Facebook/src/FacebookHandler.cs:
```diff
@@ -60,16 +58,15 @@ protected override async Task<AuthenticationTicket> CreateTicketAsync(ClaimsIden
+            var key = Encoding.ASCII.GetBytes(Options.AppSecret);
+            var tokenBytes = Encoding.ASCII.GetBytes(accessToken);
+            var hash = HMACSHA256.HashData(key, tokenBytes);
+            var builder = new StringBuilder();
```
I considered changing this to use `Convert.ToHexString().ToLowerInvariant()` too, but wasn't sure on which was "better" on a non-test code path given simplifying it would require the `ToLower*()` call. Will change if the simplicity would be preferred.

javiercn replied:
You might do all this without any allocation if you want:
* Stackalloc for the hash signature
* Use the Span based API to get the hex encoding on a span
* Use ToLowerInvariant to convert ToLower
* Create the final string from the span

 https://docs.microsoft.com/en-us/dotnet/api/system.memoryextensions.tolowerinvariant?view=net-5.0

martincostello replied:
Like this?

```csharp
var secretBytes = Encoding.ASCII.GetBytes(Options.AppSecret);
var tokenBytes = Encoding.ASCII.GetBytes(accessToken);

Span<byte> hashedBytes = stackalloc byte[32];

var written = HMACSHA256.HashData(secretBytes, tokenBytes, hashedBytes);

if (written != hashedBytes.Length)
{
    throw new InvalidOperationException("Could not compute the hash for the 'appsecret_proof' value.");
}

var hashStringUpper = Convert.ToHexString(hashedBytes).AsSpan();

Span<char> hashStringLower = stackalloc char[hashStringUpper.Length];

if (hashStringUpper.ToLowerInvariant(hashStringLower) != hashStringUpper.Length)
{
    throw new InvalidOperationException("Could not compute the hash for the 'appsecret_proof' value.");
}

return new(hashStringLower);
```

javiercn replied:
Ah, my bad. I saw the ToHexString span API and assumed it could do it without the intermediate hex string conversion. I'm surprised that there is no overload to do this without generating a string

javiercn replied:
I thought it was ToHexString(span, resultSpan)

javiercn replied:
Filed an issue on the runtime https://github.com/dotnet/runtime/issues/58937 to see if it gets added in the future

martincostello committed:
Assert on appsecret_proof

Assert that the correct secret proof is added to the request URL.

martincostello committed:
Update URL

Update URL that redirects.

martincostello commented:
While I was playing around with Javier's suggestion I found that the tests didn't explicitly assert on the `appsecret_proof` value generated by the Facebook handler, so I've added an assert for that. I also updated a link that now redirects elsewhere.

martincostello committed:
Fix invalid documentation link

Update link that resulted in a 404.

martincostello committed:
Expand TwitterHandler coverage

Add test coverage for CreateTicketAsync(), ObtainAccessTokenAsync() and RetrieveUserDetailsAsync().

Tratcher commented:
Thanks
